### PR TITLE
feat: implementation of TS common viewContent, URL fragment & remote URL content

### PR DESCRIPTION
**Pull request by micwallace**
- Add view content processing, to inject common view content when a view is loaded.
- Add URL hash/fragment location when loading the view.
- Add the ability to load tokenscript view from a remote URL.
- Add support for updated XML version namespace

### CODE REVIEW

**Comment by micwallace** [comment]
@JamesSmartCell Good news, CDATA is working in Android :-)

**Comment by hboon** [comment]
>- Add view content processing, to inject common view content when a view is loaded. ...

@micwallace would you help write a description/issue so iOS can implement it for feature parity?

File: lib/src/main/java/com/alphawallet/token/tools/TokenDefinition.java [hunk 1]
```diff
@@ -1249,8 +1257,8 @@ public String getCardData(String tag)
     {
         TSTokenView view = tokenViews.views.get("view");
 
-        if (tag.equals("view")) return view.tokenView;
-        else if (tag.equals("style")) return view.style;
+        if (tag.equals("view")) return view.getTokenView();
+        else if (tag.equals("style")) return view.getTokenView();
```
[the remark this answers is not in the document]
Comment (micwallace) [reply lib/src/main/java/com/alphawallet/token/tools/TokenDefinition.java]: Indeed! My bad, good catch thanks Seaborne. I've updated

**Comment by micwallace** [comment]
Hey @hboon ,

I've got to focus on Devcon stuff for most of this week, but there is some preliminary information here:
https://github.com/TokenScript/tokenscript-cli/pull/9

It basically allows 2 or more cards to share the same javascript/html/css.

**Comment by seabornlee** [comment]
@micwallace There is a test case failure, can you check it out?
image

File: lib/src/main/java/com/alphawallet/token/entity/TSTokenView.java [hunk 2]
```diff
@@ -7,18 +7,36 @@
...
+    public TSTokenView(Element element, TokenDefinition tokenDef)
     {
+        this.element = element;
+        this.tokenDef = tokenDef;
+    }
+
+    private void generateTokenView(Element element)
+    {
+
+        if (!Objects.equals(this.getUrl(), "")){
```
Comment (JamesSmartCell) [inline lib/src/main/java/com/alphawallet/token/entity/TSTokenView.java:36]: ```{``` should be on new line

File: lib/src/main/java/com/alphawallet/token/entity/TSTokenView.java [hunk 3]
```diff
@@ -51,8 +74,36 @@ public TSTokenView(Element element)
...
 
-        tokenView = lView;
-        style = lStyle;
+        tokenView += lView;
+        style += lStyle;
+    }
+
+    public String getTokenView()
+    {
+        if (tokenView.isEmpty()){
```
Comment (JamesSmartCell) [inline lib/src/main/java/com/alphawallet/token/entity/TSTokenView.java:83]: ```{``` should be on new line

File: lib/src/main/java/com/alphawallet/token/entity/TSTokenView.java [hunk 4]
```diff
@@ -51,8 +74,36 @@ public TSTokenView(Element element)
...
+            generateTokenView(this.element);
+        }
+
+        return tokenView;
+    }
+
+    public String getStyle()
+    {
+        if (style.isEmpty()){
```
Comment (JamesSmartCell) [inline lib/src/main/java/com/alphawallet/token/entity/TSTokenView.java:92]: ```{``` should be on new line

File: lib/src/main/java/com/alphawallet/token/tools/TokenDefinition.java [hunk 5]
```diff
@@ -670,7 +680,7 @@ private void processTokenCardElements(Element card) throws Exception
                     break;
                 case "view": //TODO: Localisation
                 case "item-view":
-                    TSTokenView v = new TSTokenView(element);
+                    TSTokenView v = new TSTokenView(element, this);
```
Comment (JamesSmartCell) [inline lib/src/main/java/com/alphawallet/token/tools/TokenDefinition.java:683]: we probably should do this with an interface, but it's ok for now.

File: lib/src/main/java/com/alphawallet/token/tools/TokenDefinition.java [hunk 6]
```diff
@@ -767,6 +777,33 @@ else if (thisDate.before(schemaDate))
         }
     }
 
+    public boolean isSchemaLessThanMinimum()
+    {
+
+        if (nameSpace == null)
```
Comment (JamesSmartCell) [inline lib/src/main/java/com/alphawallet/token/tools/TokenDefinition.java:783]: Code safety - wrap in braces.

File: app/src/main/java/com/alphawallet/app/ui/FunctionActivity.java [hunk 7]
```diff
@@ -166,12 +167,16 @@ private void displayFunction(String tokenAttrs)
             TSAction action = functions.get(actionMethod);
             String magicValues = viewModel.getAssetDefinitionService().getMagicValuesForInjection(token.tokenInfo.chainId);
 
-            String injectedView = tokenView.injectWeb3TokenInit(action.view.tokenView, tokenAttrs, tokenId);
-            injectedView = tokenView.injectJSAtEnd(injectedView, magicValues);
-            injectedView = tokenView.injectStyleAndWrapper(injectedView, action.style + "\n" + action.view.style);
+            if (Objects.equals(action.view.getUrl(), "")){
```
Comment (JamesSmartCell) [inline app/src/main/java/com/alphawallet/app/ui/FunctionActivity.java:170]: ```{``` newline

**Review by JamesSmartCell — COMMENTED** [review]
The code style needs to be normalised a little (this is a formal language), but it's all good other than that.

**Review by JamesSmartCell — APPROVED** [review]

**Pull request merged**